### ISSUE TITLE
Local variable becomes static when passing thru 2 functions

### ISSUE THREAD

**Issue by Telecs** [Scirra/Construct-3-bugs#3519]
## Problem description

Hello,
As the title of the issue states, I noticed a problem when I call a function thru another one. it seems that if I increment a variable thru the call of a second function, the variable suddenly turns to be static and keeps the value tick after tick.

## Attach a .c3p

Here is a link to [a simple example](https://www.dropbox.com/s/vwaov9i4gklltjy/test%20local%20variable.c3p?dl=0).

## Steps to reproduce

1. Run the game and open the console to check the value of the variable displayed each second.

## Observed result

The value of the variable is continuously incremented

## Expected result

The variable should reset each frame and the log should display 1 each time I call the first function.

## More details

I tested this problem only directly in the Construct previsualisation editor so I don't have any more info.

**Affected browsers/platforms:** 

I have no clue, sorry.

**First affected release:** 

No clue either.

## System details

<details><summary>View details</summary>

Platform information
Browser: NW.js
Browser version: 0.33.3
Browser engine: Chromium
Browser architecture: 64-bit
Context: nwjs
Operating system: Windows
Operating system version: 10
Operating system architecture: 64-bit
Device type: desktop
Device pixel ratio: 1
Logical CPU cores: 4
Approx. device memory: 8 GB
User agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (@8920e690dd011895672947112477d10d5c8afb09) NWjs/0.33.3 (like Chrome/69.0.3497.100)
C3 release: r171.3 (stable)
Language setting: en-US

Local storage
Storage quota (approx): 44 gb
Storage usage (approx): 531 mb (1.2%)
Persistant storage: Yes

Browser support notes
This list contains missing features that are not required, but could improve performance or user experience if supported.

Nothing is missing. Everything is OK!
WebGL information
Version string: WebGL 2.0 (OpenGL ES 3.0 Chromium)
Numeric version: 2
Supports NPOT textures: yes
Supports GPU profiling: no
Supports highp precision: yes
Vendor: Google Inc.
Renderer: ANGLE (Radeon(TM) RX 460 Graphics Direct3D11 vs_5_0 ps_5_0)
Major performance caveat: no
Maximum texture size: 16384
Point size range: 1 to 1024
Extensions:

EXT_color_buffer_float
EXT_texture_filter_anisotropic
OES_texture_float_linear
WEBGL_compressed_texture_s3tc
WEBGL_compressed_texture_s3tc_srgb
WEBGL_debug_renderer_info
WEBGL_debug_shaders
WEBGL_lose_context
Audio information
System sample rate: 48000 Hz
Output channels: 2
Output interpretation: speakers
Supported decode formats:

Ogg Opus (audio/ogg; codecs=opus)
WebM Opus (audio/webm; codecs=opus)
Ogg Vorbis (audio/ogg; codecs=vorbis)
WebM Vorbis (audio/webm; codecs=vorbis)
MPEG-4 AAC (audio/mp4; codecs=mp4a.40.5)
MP3 (audio/mpeg)
FLAC (audio/flac)
PCM WAV (audio/wav; codecs=1)
Supported encode formats:

WebM Opus (audio/webm; codecs=opus)
Video information
Supported decode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)
Ogg Theora (video/ogg; codecs=theora)
H.264 (video/mp4; codecs=avc1.42E01E)
Supported encode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)

</details>

**Comment by shortercode** [comment]
This seems to be a somewhat strange edge case in variable scoping, I'm assigning to @AshleyScirra to review.

JS equivalent
```javascript

let fn1;
let fn2;

{// group
    let counter = 0;
    fn1 = () => {
        counter++;
        console.log(counter);
    }
    fn2= () => fn1();
}
```

Calling "fn2" results in an incrementing value, but calling "fn1" just emits 1 each time.

**Comment by AshleyScirra** [comment]
Construct has a mechanism which basically copies local variables in nested function calls to give them a new unique value. This is important for recursive functions so local variables work independently within each call. The way Construct does it isn't really analogous to any programming languages (it only takes effect within nested calls, hence the indirect call in the provided project) - it's a bit of an odd design really, but was implemented years ago for the old legacy Function plugin in C2 and has made it all the way to the new C3 builtin functions for backwards compatibility. Anyway I've hacked in a fix, but we could probably do with rethinking how this part of C3 works at some point...
